### PR TITLE
Try: Manage focus through EditableProvider

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to explore options for removing the need for block implementers to manage focus state of a block, often by passing through `focus` and `setFocus` props to the Editable component.

__Implementation notes:__

It builds upon the [`EditableProvider`](https://github.com/WordPress/gutenberg/blob/master/blocks/editable/provider.js) introduced in #1943, which uses [React context](https://facebook.github.io/react/docs/context.html) to communicate transparently between the editor infrastructure and Editable instances.

This is made more difficult because it requires bidirectional communication (Editable `onFocus` -> Editor, Editor `focus` -> Editable), and detecting changes in context values is not accounted in components implementing `shouldComponentUpdate` (https://github.com/facebook/react/issues/2517). Taking cues from a project [react-side-context](https://github.com/Morhaus/react-side-context), the workaround here is to create an event emitter in state which detects changes and emits a `change` event to incur a forced update.

In retrospect, I found that many blocks are making a conscious choice to use `focus` to determine whether block inspector or toolbar controls should be rendered. In most cases, they are primarily concerned that it is truthy ("is focussed"), and in a few cases are managing disparate focus states (quote focused in text vs. caption). The latter of these requires `setFocus`, whereas the former does not. In any case, there's not as much benefit to be gained here as I'd hoped, aside from simple cases where focus is managed outside the view of the block implementer. It could be argued that explicitness is actually favorable here, both to clarify actual behavior, and expose block implementers to available features.

This all said, I'm not sure if this is something we ought to move on, but thought I'd publish all the same to invite discussion on ideas of using context or other patterns to remove concerns from the responsibility of block implementers.

### CODE REVIEW

**Comment by youknowriad** [comment]
I was a bit sceptical about this but the fact that it's opt-in is a good thing IMO. If a block decides to explicitly handle the focus, it should opt-in (using a specific prop maybe). 

I think if we follow this approach, we should try to avoid the necessity of the `focus` check before rendering the toolbars (Could be a CSS trick).

The event emitter bothers a bit though, it's hard to follow. I wonder if we can't replace it with `connecting` the Editable directly

**Comment by iseulde** [comment]
To me, requiring blocks to use `focus` for something like the inspector feels strange. Will there ever be a case were the block should have something in the inspector when it is not focussed? IMHO, inspector and toolbar should be separate from the edit function.

This approach would also make it easier for #2296. In order to move the logic to editable, it needs to be aware of the previous and last block. With the EditableProvider I wouldn't have to pass these through all the block components.

**Comment by aduth** [comment]
@iseulde These are great points I totally overlooked. If we adopted something like this, it'd be possible in most cases for block implementers to just render the block controls and not worry about focus (also prevents from implementers forgetting to include this check).

Now, whether block controls are in fact separate from `edit`? I'm not sure, since they're certainly relevant only to the editing experience, not the block as a standalone thing. Where else might you imagine they live? As we had previously with a `controls` property?

**Comment by aduth** [comment]
A challenge here is knowing how to associate the Fill of individual blocks with the Slot for toolbar controls, since the slot is "global". One option might be to assign a unique name for the toolbar slot of each block (maybe using its `uid`), but this would require that the block's toolbar have awareness of the block ID. Is this passed by block implementers, or continue to abuse context to pass this information in implicitly?

**Comment by aduth** [comment]
Going to take a different approach here.
